### ISSUE TITLE
[BUG] Strange conversion to camel case

### ISSUE THREAD

**Issue by judgej** [OpenAPITools/openapi-generator#6559]
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] What's the version of OpenAPI Generator used?
- [ ] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

Attribute names that include upper case letters are not being converted to camel case correctly.
This affectes model attributes, model names etc.

This can be demostrated through a lambda in a mustache template:

```text
1. id {{#lambda.camelcase}}id{{/lambda.camelcase}}
2. ID {{#lambda.camelcase}}ID{{/lambda.camelcase}}
3. id_more {{#lambda.camelcase}}id_more{{/lambda.camelcase}}
4. ID_MORE {{#lambda.camelcase}}ID_MORE{{/lambda.camelcase}}
```

This results in:

```text
1. id id
2. ID iD
3. id_more idMore
4. ID_MORE iDMORE
```

I believe 2 and 4 are wrong. For 2, the upper-case token should be treated as one word, so should be converted to lower case: "id".

For 4, there are two words, but of which should be converted to lower case, then concatenated with the second work capitalised: "idMore".

The model data supplied to the templates does the same conversion to an "ID" in the OpenAPI description, giving this:

```json
{
      "openApiType" : "integer",
      "baseName" : "ID",  <<<
      "getter" : "getID",  <<<
      "setter" : "setID",  <<<
      "dataType" : "int",
      "datatypeWithEnum" : "int",
      "dataFormat" : "int32",
      "name" : "iD",  <<<
      ...
}
```

Here I would expect to see "ID", "getId", "setId" and "id".

##### openapi-generator version

The latest CLI docker container `openapitools/openapi-generator-cli`

`5.0.0-SNAPSHOT`

##### OpenAPI declaration file content or url

This shows the same effect, in any mustache template you like:

```text
1. id {{#lambda.camelcase}}id{{/lambda.camelcase}}
2. ID {{#lambda.camelcase}}ID{{/lambda.camelcase}}
3. id_more {{#lambda.camelcase}}id_more{{/lambda.camelcase}}
4. ID_MORE {{#lambda.camelcase}}ID_MORE{{/lambda.camelcase}}
```

Or any OpenAPI description with an UPPERCASE property name.

##### Command line used for generation

    docker run --rm openapitools/openapi-generator-cli ...

##### Steps to reproduce

See above.

##### Related issues/PRs

Not been able to find any.

##### Suggest a fix

I'm not a Java guy, and have tried and failed to set up an environment in the past, so can do nothing to help.

**Comment by judgej** [comment]
You may disagree that "ID" should become "id" as camelcase, but "iD" really looks silly, and we have no control over third-party APIs, so cannot simply change the properties to lowercase.

**Comment by judgej** [comment]
Here is a  =nice convertert site that does it correctly IMO (does Pascal case, but does correctly what I am suggesting is wrong here:

https://en.toolpage.org/tool/camelcase

Try:

```text
id
ID
id_more
ID_MORE
```

gives:

```text
Id
Id
IdMore
IdMore
```

**Comment by judgej** [comment]
There is plenty of advice on SO on how abbreviations are handled, and some contradictory. Some say two-letter abbreviations should be left upper case (e.g. `ID`) and some not (e.g. `Id`) but none - absolutely none - give `iD` as a good example of handling an abbreviation in camel case.

This areticle has some good examples:

https://www.approxion.com/capital-offenses-how-to-handle-abbreviations-in-camelcase/

This is the statement I personally aggree with:

   The only convention that works reliably is this: “Treat abbreviations just like ordinary words”.

So `My_ID` becomes `myId`. `UNESCO_ID` is then `unescoId`.

**Comment by blue00** [comment]
I have the same problem with the typescript-angular converter using the option `"modelPropertyNaming": "camelCase"`:
"VAT" becomes "vAT", but I expected it to become "vat".

**Comment by judgej** [comment]
I found recently that it is possible to provide parameters to override the mappings for constants. So you you find a constant name (from a list of status values, for example) clashes with a reserved word, you can override it in the generated code.

I'm wondering if there is some kind of mapping option for other properties (and not just emum values) so you could tell the code generator to hande "VAT" as "vat" in its internal usage (to generate methods, properties etc.) while keeping the correct letter case for serialising and deserialising. That may offer a ~~solution~~ *workaround* to the issue.

**Comment by judgej** [comment]
Is there a way to import a custom camel case lambda at runtime, and using that to produce desired conversion?

**Comment by bradydan** [comment]
Update from Alex at PODWW:

> I completely overlooked filtering by date, it’s definitely a feature we will be implementing!

**Comment by abraha2d** [comment]
Did some digging. This used to work perfectly, but broke with v4.3.0 (v4.2.3 is the last version that converted SCREAMING_SNAKE_CASE to camelCase "properly").

**Comment by judgej** [comment]
Love the name screaming camel case :-)

My issue with the conversion is that it does not take account of real life human-oriented design. People automatically use ID and VAT, and other common abbreviations in upper case, because it just makes sense. Coders want to see consistent camel case. The code generator is the thing that has to bring these two worlds together.

Even if there were a way of "correcting" properties in config to treat thingID as thingId, I would be happy. Maybe an external filter to transform the api description is the solution?

**Comment by ghe-bm** [comment]
Is it planned to fix the camel case behavior?

**Comment by judgej** [comment]
I don't think it is going to be changed. If nothing else, it could be a breaking change for a large number of clients.

The approach that should probably be encouraged is to never use upper-case abbreviations such as VAT and ID in an Open API description.

Maybe there is scope to be able to define aliases for names to be used for all internal references within a generated client, while keeping the original name just for use on-the-wire (generating and parsing JSON and XML parameters). There is an "alias" parameter to do something like this, but I've not managed to get it to work consistently. However, that was some time ago, and the functionality of aliases may have been improved since then. In the meantime, we just put put up with the likes of `vATNumber` and `iD` as parameters and methods.

Personally I think a manual alias or mapping list will be the way to go. Trying to camel-case `VATNumber` into `vatNumber` with an algorithm that does not have strange edge-cases is a hard problem. So a developer looking at what is generated, and having the ability to "fix it" manually based on knowledge of what was intended is going to be the most reliable method. It also won't introduce breaking changes unexpectedly.

**Comment by wing328** [comment]
FYI. we added some options to allow users to map property/parameter/model/enum name to whatever they want to address issues like this (as a workaround): https://github.com/openapitools/openapi-generator/blob/master/docs/customization.md#name-mapping

I'll see later what I can do to improve `camelize` to handle words starting with capital letters.

**Comment by judgej** [comment]
Very nice @wing328 Does this mapping effectuvely sit in the camelisation flow, so instead of being give the default camel mapping, you can put your own custom mapping in, in its place?

**Comment by wing328** [comment]
it will skip any change to the input and directly provide the mapping given by the user. you can give it a try following the example in the doc.
